HTTP2 remove window between client stream shutdown and stream count

shinrich opened the pull request:
Follow up to PR #2155.  @maskit and @masaori335 pointed out in that PR, that under load there is a window of opportunity for the client to send a new stream request after a previous stream close.  From the client perspective, it is keeping under the max_concurrent limit, but there was a window of opportunity for the client new stream request to come in before ATS had decremented the current stream count, so ATS would incorrectly send back a max concurrent error.

The original code was rescheduling to do the final stream cleanup to ensure that the stream we are destroying in not the stack.  I changed this to do reference counting and delete on exit from the event handler if the stream was marked to be terminated.  This is the same technique used for HttpSM.  By using this approach there is no scheduling window that would allow another client stream request to sneak in before the current stream is fully destroyed.

I've done the nghttp test that @maskit identified in PR #2155 (we should bring that into an autest).  I have also run it in our production environment against a mega-emoji page that has been causing us problems in firefox.  I'm now running it in general production to check for the possibility that this change introduced leaks or shutdown crashes.

maskit commented:
We already have:
```
bool delete_stream(Http2Stream *stream);
void release_stream(Http2Stream *stream);
```
and now we are adding another one... We need a documentation (or comments) about stream / session closing sequence.

masaori335 commented on proxy/http2/Http2Stream.h:
```diff
@@ -249,6 +251,8 @@ class Http2Stream : public ProxyClientTransaction
   ChunkedHandler chunked_handler;
   Event *cross_thread_event = nullptr;
+  int reentrancy_count      = 0;
+  bool terminate_stream     = false;
```
`closed` flag is not enough? Really need one more new flag?

SolidWallOfCode replied:
It might be possible to multi-value the `closed` flag, but fundamentally you have 3 states that must be tracked.

* open
* should be closed / prospective close.
* actually closed.

The middle state is the point of this PR, to mark the intent to close when the stream cannot yet be safely closed (waiting for stack unwind).

zwoop commented on proxy/http2/Http2Stream.cc:
```diff
@@ -33,6 +33,8 @@ Http2Stream::main_event_handler(int event, void *edata)
 {
   Event *e = static_cast<Event *>(edata);
 
+  reentrancy_count++;
```
Nice color on this bike shed!!

shinrich commented:
Addressed some of @masaori335 and @maskit comments. 

Got rid of sent_delete flag.  But I think we need a terminate_stream flag.  I added a discussion of the flags in Http2Stream.h.  Got rid of kill_stream.  Sank that logic into destroy.

zwoop commented:
This has been back ported to 7.1.0.
